### PR TITLE
Ensure logs bridge API doesn't contain implementation details

### PR DESCRIPTION
**Pull request by jack-berg**
Contributes to #3268.

@MrAlias did some good work in the metrics API / SDK recently in #3171 and #3067 to ensure that the metrics API spec doesn't contain SDK implementation details.

This PR adopts similar language in the Logs Bridge API / SDK documents, which includes breaking out a `noop.md` document.

### CODE REVIEW

File: specification/metrics/sdk.md [hunk 1]
```diff
@@ -104,10 +104,6 @@ working Meter MUST be returned as a fallback rather than returning null or
 throwing an exception, its `name` SHOULD keep the original invalid value, and a
 message reporting that the specified value is invalid SHOULD be logged.
 
-When a Schema URL is passed as an argument when creating a `Meter` the emitted
-telemetry for that `Meter` MUST be associated with the Schema URL, provided
-that the emitted data format is capable of representing such association.
-
```
Comment (jack-berg) [inline specification/metrics/sdk.md:110]: This line is redundant to the first paragraph in this section, which says that schema URL is included in instrumentation scope.

File: specification/logs/bridge-api.md [hunk 2]
```diff
@@ -57,13 +57,6 @@ Normally, the LoggerProvider is expected to be accessed from a central place.
 Thus, the API SHOULD provide a way to set/register and access a global default
 LoggerProvider.
 
-Notwithstanding any global LoggerProvider, some applications may want to or have
-to use multiple LoggerProvider instances, e.g. to have different configuration
-(like [LogRecordProcessors](sdk.md#logrecordprocessor)) for each (and
-consequently for the Loggers obtained from them), or because it's easier with
-dependency injection frameworks. Thus, implementations of LoggerProvider SHOULD
-allow creating an arbitrary number of instances.
```
Comment (tigrannajaryan) [inline specification/logs/bridge-api.md:65]: This is mirroring the working of [trace api](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md). Do we think this is unnecessary or it should be moved to sdk doc?

File: specification/logs/bridge-api.md [hunk 3]
```diff
@@ -74,68 +67,53 @@ The LoggerProvider MUST provide the following functions:
 
 This API MUST accept the following parameters:
 
-- `name` (required): This name SHOULD uniquely identify the [instrumentation scope](../glossary.md#instrumentation-scope),
```
Comment (tigrannajaryan) [inline specification/logs/bridge-api.md:77]: I find "required" and "optional" more concise than the "can provide, but it is up to their discretion, therefore ... needs to accept ... but MUST NOT" lengthy wording that seems to replace this. Why do we need this change?

File: specification/logs/bridge-api.md [hunk 4]
```diff
@@ -74,68 +67,53 @@ The LoggerProvider MUST provide the following functions:
...
-work with an outdated configuration or by ensuring that new configuration
-applies also to previously returned Loggers.
-
-Note: This could, for example, be implemented by storing any mutable
-configuration in the `LoggerProvider` and having `Logger` implementation objects
-have a reference to the `LoggerProvider` from which they were obtained.
-If configuration must be stored per-Logger (such as disabling a certain `Logger`),
-the `Logger` could, for example, do a look-up with its name+version+schema_url+include_trace_context+attributes
-in a map in the `LoggerProvider`, or the `LoggerProvider` could maintain a registry
-of all returned `Logger`s and actively update their configuration if it changes.
```
Comment (tigrannajaryan) [inline specification/logs/bridge-api.md:126]: This is important isn't it? I can't see where this went. It defines an API behavior too. So perhaps get rid of "for example part", but I think we can't just delete it all.

File: specification/logs/bridge-api.md [hunk 5]
```diff
@@ -74,68 +67,53 @@ The LoggerProvider MUST provide the following functions:
...
-
-The effect of associating a Schema URL with a `Logger` MUST be that the telemetry
-emitted using the `Logger` will be associated with the Schema URL, provided that
-the emitted data format is capable of representing such association.
```
Comment (tigrannajaryan) [inline specification/logs/bridge-api.md:130]: This is important too. I am not sure why this is deleted. This is an observable effect of the system caused by an API call. It is part of the API specification.

File: specification/logs/bridge-api.md [hunk 6]
```diff
@@ -167,24 +145,28 @@ fields:
 - [Body](./data-model.md#field-body)
 - [Attributes](./data-model.md#field-attributes)
 
+User can provide each of these fields, but it is up to their discretion.
+Therefore, this API needs to be structured to accept these fields, but MUST NOT
+obligate a user to provide them.
```
Comment (tigrannajaryan) [inline specification/logs/bridge-api.md:150]: Perhaps we can say more concisely "All fields are optional" instead?

File: specification/logs/noop.md [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+computation overhead and eliminate OpenTelemetry related output. For
+this reason, the `LoggerProvider` MUST NOT return a non-empty error or log
+any message for any operations it performs.
+
+All operations a `LoggerProvider` provides MUST be safe to be run
+concurrently.
+
+### Logger Creation
+
+[New Logger instances are always created with a LoggerProvider](./bridge-api.md#loggerprovider).
```
Comment (tigrannajaryan) [inline specification/logs/noop.md:52]:
```suggestion
New `Logger` instances are always created with a [LoggerProvider](./bridge-api.md#loggerprovider).
```

File: specification/logs/noop.md [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+
+[The API specifies multiple parameters](./bridge-api.md#loggerprovider) that
+need to be accepted by the creation operation. The `LoggerProvider` MUST
+accept these parameters. However, the `LoggerProvider` MUST NOT validate
+any argument it receives.
+
+## Logger
+
+A `Logger` is always created by a `LoggerProvider`. The No-Op MUST NOT provide
+a way for a user to create a `Logger` other than by a No-Op `LoggerProvider`.
```
Comment (tigrannajaryan) [inline specification/logs/noop.md:67]: >The No-Op MUST NOT provide
>a way for a user to create a `Logger` other than by a No-Op `LoggerProvider`.

I don't understand the purpose of this part.

File: specification/logs/noop.md [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+
+The `Logger`s created by the No-Op need to hold as small a memory
+footprint as possible. Therefore, all `Logger`s created MUST NOT hold
+configuration or operational state.
+
+The `Logger` MUST NOT return a non-empty error or log any message for any
+operations it performs.
+
+All operations a `Logger` provides MUST be safe to be run concurrently.
```
Comment (tigrannajaryan) [inline specification/logs/noop.md:76]: This should be part of Logger's API spec. This behavior should not be different between noop and regular implementations.

File: specification/logs/noop.md [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+
+### Emit LogRecord
+
+The No-Op `Logger` MUST allow
+for [emitting LogRecords](./bridge-api.md#emit-logrecord).
+
+[The API specifies multiple parameters](./bridge-api.md#emit-logrecord) that
+need to be accepted by `emit`. The `Logger` MUST accept these parameters.
+However, the `Logger` MUST NOT validate any argument it receives.
```
Comment (tigrannajaryan) [inline specification/logs/noop.md:85]: We have a similar sentence above. To avoid repetition maybe we replace it by general wording that say "calls to noop MUST NOT validate their arguments"?

File: specification/logs/sdk.md [hunk 11]
```diff
@@ -33,28 +35,48 @@
...
+
+The SDK SHOULD allow the creation of multiple independent `LoggerProviders`s.
 
 ### Logger Creation
 
 New `Logger` instances are always created through a `LoggerProvider`
-(see [Bridge API](bridge-api.md)). The `name`, `version` (optional), and `attributes` (optional)
-supplied to the `LoggerProvider` must be used to create
+(see [Bridge API](bridge-api.md)). The `name`, `version` (optional),
+`schema_url` (optional), and `attributes` (optional) supplied to
```
Comment (tigrannajaryan) [inline specification/logs/sdk.md:59]: Why is schema_url deleted?

File: specification/logs/sdk.md [hunk 12]
```diff
@@ -33,28 +35,48 @@
...
 an [`InstrumentationScope`](../glossary.md#instrumentation-scope) instance which
 is stored on the created `Logger`.
 
+In the case where an invalid `name` (null or empty string) is specified, a
+working `Logger` MUST be returned as a fallback rather than returning null or
+throwing an exception, its `name` SHOULD keep the original invalid value, and a
+message reporting that the specified value is invalid SHOULD be logged.
+
+If the case where `include_trace_context` is not specified, its default value
+SHOULD be true by default.
```
Comment (tigrannajaryan) [inline specification/logs/sdk.md:70]: I think this is already in the API spec and is not needed here.

File: specification/logs/sdk.md [hunk 13]
```diff
@@ -99,6 +121,15 @@ decide if they want to make the flush timeout configurable.
 `ForceFlush` MUST invoke `ForceFlush` on all
 registered [LogRecordProcessors](#logrecordprocessor).
 
+## Logger
+
+If a `Logger` was obtained with `include_trace_context=true`, the `LogRecord`s
+it [emits](./bridge-api.md#emit-logrecord) MUST automatically include the Trace
+Context from the active Context, if Context has not been explicitly set.
```
Comment (tigrannajaryan) [inline specification/logs/sdk.md:128]: >if Context has not been explicitly set.

What does this refer to? Do we have a way to set the context explicitly?

**Review by tigrannajaryan — COMMENTED** [review]
Thanks @jack-berg

File: specification/logs/sdk.md (hunk 13 above)
Comment (jack-berg) [reply specification/logs/sdk.md:128]: Yes. There's a section on [explicit context injection](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/bridge-api.md#explicit-context-injection). Also, the definition of [LogRecord](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/bridge-api.md#logrecord) which is accepted by `emit(LogRecord)` includes a `Context` argument.

File: specification/logs/sdk.md (hunk 12 above)
Comment (jack-berg) [reply specification/logs/sdk.md:70]: I removed the details about the default value from the API spec since its an implementation concern. 

Trying to follow the pattern established in #3171 to try not to let SDK implementation details leak into the API.

File: specification/logs/sdk.md (hunk 11 above)
Comment (jack-berg) [reply specification/logs/sdk.md:59]: `schema_url` is missing the current version. Maybe an accidental omission? I've added it in because its part of the scope.

File: specification/logs/noop.md (hunk 10 above)
Comment (jack-berg) [reply specification/logs/noop.md:85]: That seems reasonable to me. For context, I was trying to stay consistent with the newly added [metric noop implementation](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/noop.md) spec, where each operation repeats that it must not perform validation.

File: specification/logs/bridge-api.md (hunk 6 above)
Comment (jack-berg) [reply specification/logs/bridge-api.md:150]: When similar changes were being proposed to the metrics spec, I commented with the same opinion [here](https://github.com/open-telemetry/opentelemetry-specification/pull/3074#discussion_r1074017566). @MrAlias convinced me that "optional" is overly simplistic, and doesn't make it clear whether they're optional to implement or optional for the user to provide. 

I'm in favor of consistent language more than anything else.

File: specification/logs/bridge-api.md (hunk 3 above)
Comment (jack-berg) [reply specification/logs/bridge-api.md:77]: I think my reply [here](https://github.com/open-telemetry/opentelemetry-specification/pull/3275/files#r1125023786) is equally applicable. Can we apply whatever resolution we come to in that comment to this conversation as well?

File: specification/logs/bridge-api.md (hunk 2 above)
Comment (jack-berg) [reply specification/logs/bridge-api.md:65]: I [moved it to the SDK](https://github.com/open-telemetry/opentelemetry-specification/pull/3275/files#diff-fbe59ced3c46267bf03f50f8330ec1f38f56a668fbeb88bb564c63bb47524c60R53) in a simpler form which [mirrors the more recent language of the metrics SDK](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#meterprovider-creation). 

I do think we should update the trace documents to be consistent.

File: specification/logs/bridge-api.md (hunk 4 above)
Comment (jack-berg) [reply specification/logs/bridge-api.md:126]: Was trying to be consistent with the same removal in [#3171](https://github.com/open-telemetry/opentelemetry-specification/pull/3171/files#diff-055661289c337a191d4767d311dbdf1fe63e0aa07f9d58944da92c54f4382df1L176-L183). Not sure what the motivation behind removing it from the metrics API spec was. Maybe @MrAlias remembers? Naively, I might guess that it was removed because the focuses too much on specific implementation details rather than expected behavior.

File: specification/logs/bridge-api.md (hunk 5 above)
Comment (jack-berg) [reply specification/logs/bridge-api.md:130]: Isn't that implied by `schema_url` being part of Logger identity? We don't include similar language about the other scope fields (i.e. name, version, attributes).

File: specification/logs/bridge-api.md (hunk 3 above)
Comment (tigrannajaryan) [reply specification/logs/bridge-api.md:73]: I would prefer that we define in one place that "optional" means "can provide, but it is up to their discretion, therefore ... needs to accept ... but MUST NOT" instead of repeating it every time we have an optional parameter.

File: specification/logs/bridge-api.md (hunk 5 above)
Comment (tigrannajaryan) [reply specification/logs/bridge-api.md:126]: I don't think it is implied by the identity. It may be part of the identity but never be included on the wire. The following phrasing mandates explicitly the the schema URL should be on the wire and is not just an identifying field:
>the telemetry emitted using the `Logger` will be associated with the Schema URL

I believe this needs to stay. Deleting it changes the spec meaningfully and that change is wrong, IMO.

File: specification/logs/noop.md (hunk 8 above)
Comment (jack-berg) [reply specification/logs/noop.md:64]: Looks like the metrics noop document ended up feeling the same way and is removing this in #3322.

Will remove this extra language.

File: specification/logs/sdk.md (hunk 12 above)
Comment (jack-berg) [reply specification/logs/sdk.md:70]: After thinking about it a bit more, I agree. The other optional arguments have a null or empty default value. This parameter is different because its a boolean and it doesn't make sense to have a null boolean here. The user may choose not to provide it because they desire the default behavior, which is `include_trace_context=true`.

File: specification/logs/noop.md (hunk 10 above)
Comment (jack-berg) [reply specification/logs/noop.md:80]: Fair enough. Will remove. Will leave the noop metrics doc along for now and propose the simplification in a followup.

File: specification/logs/bridge-api.md (hunk 3 above)
Comment (jack-berg) [reply specification/logs/bridge-api.md:73]: Added a section defining optional and required parameters.

File: specification/logs/bridge-api.md (hunk 4 above)
Comment (jack-berg) [reply specification/logs/bridge-api.md]: Will mark as resolved as this does seem too oriented towards implementation. Please re-open it if needed.

**Comment by tigrannajaryan** [comment]
@open-telemetry/specs-logs-approvers please review.

File: specification/logs/sdk.md [hunk 14]
```diff
@@ -99,6 +118,15 @@ decide if they want to make the flush timeout configurable.
...
```
Comment (tigrannajaryan) [inline specification/logs/sdk.md:125]: I think this is a partial duplicate of "Implicit Context Injection" section in the API. Should we delete one?

File: specification/metrics/api.md [hunk 15]
```diff
@@ -156,9 +156,6 @@ at least one identifying field has a different value.
 
 The meter is responsible for creating [Instruments](#instrument).
 
-Note: `Meter` SHOULD NOT be responsible for the configuration. This should be
-the responsibility of the `MeterProvider` instead.
```
Comment (tigrannajaryan) [inline specification/metrics/api.md:160]: Should metric changes be done in a separate PR?

File: specification/logs/sdk.md (hunk 14 above)
Comment (jack-berg) [reply specification/logs/sdk.md:125]: They do overlap but the Implicit / Explicit context injection sections in the Bridge API document are targeted at users, informing them how to use the API. They don't contain any normative language. This text is targeted at SDK authors and contains normative language, so I think both are necessary. Maybe the Implicit / Explicit context injection sections could be trimmed down? But I'm not sure how.

File: specification/logs/bridge-api.md [hunk 16]
```diff
@@ -70,57 +65,39 @@ The `LoggerProvider` MUST provide the following functions:
...
-name within that library or application. In case an invalid name
-(null or empty string) is specified, a working Logger implementation MUST be
-returned as a fallback rather than returning null or throwing an exception, its
-`name` property SHOULD be set to an empty string, and a message reporting that
-the specified value is invalid SHOULD be logged. A library implementing the
-OpenTelemetry API may also ignore this name and return a default instance for
-all calls, if it does not support "named" functionality (e.g. an implementation
-which is not even observability-related). A `LoggerProvider` could also return a
-no-op `Logger` here if application owners configure the SDK to suppress telemetry
-produced by this library.
```
[the remark this answers is not in the document]
Comment (jack-berg) [reply specification/logs/bridge-api.md:89]: The following text has been moved the SDK:
> In the case where an invalid `name` (null or empty string) is specified, a
working `Logger` MUST be returned as a fallback rather than returning null or
throwing an exception, its `name` SHOULD keep the original invalid value, and a
message reporting that the specified value is invalid SHOULD be logged.

The remaining text (see below) has been removed. I think it's ok to remove this as it provides speculative guidance to alternative implementations that seems unnecessary. If the API doesn't explicitly define what must happen, it should be assumed that alternative implementations can do whatever they want. Also, the part about "suppress telemetry" doesn't exist in the SDK and is likely to be confusing to users.
>A library implementing the
OpenTelemetry API may also ignore this name and return a default instance for
all calls, if it does not support "named" functionality (e.g. an implementation
which is not even observability-related). A `LoggerProvider` could also return a
no-op `Logger` here if application owners configure the SDK to suppress telemetry
produced by this library.

File: specification/logs/bridge-api.md [hunk 17]
```diff
@@ -70,57 +65,39 @@ The `LoggerProvider` MUST provide the following functions:
...
-work with an outdated configuration or by ensuring that new configuration
-applies also to previously returned `Logger`s.
-
-Note: This could, for example, be implemented by storing any mutable
-configuration in the `LoggerProvider` and having `Logger` implementation objects
-have a reference to the `LoggerProvider` from which they were obtained.
-If configuration must be stored per-Logger (such as disabling a certain `Logger`),
-the `Logger` could, for example, do a look-up with its name+version+schema_url+include_trace_context+attributes
-in a map in the `LoggerProvider`, or the `LoggerProvider` could maintain a registry
-of all returned `Logger`s and actively update their configuration if it changes.
```
[the remark this answers is not in the document]
Comment (jack-berg) [reply specification/logs/bridge-api.md:122]: The first paragraph is covered in the SDK:
> The `LoggerProvider` MAY provide methods to update the configuration. If
configuration is updated (e.g., adding a `LogRecordProcessor`), the updated
configuration MUST also apply to all already returned `Logger`s (i.e. it MUST
NOT matter whether a `Logger` was obtained from the `LoggerProvider` before or
after the configuration change). Note: Implementation-wise, this could mean
that `Logger` instances have a reference to their `LoggerProvider` and access
configuration only via this reference.

The second paragraph was removed in the metrics documents. It could stay, but IMO doesn't provide much value. It also references "disabled a certain `Logger`", a feature which doesn't currently exist.

File: specification/logs/bridge-api.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (jack-berg) [reply specification/logs/bridge-api.md:89]: The motivation for this was for alignment with metrics, and in this case I agree with the changes.

**Review by tigrannajaryan — APPROVED** [review]

**Review by reyang — APPROVED** [review]
